rollout.go

```diff
@@ -29,10 +37,46 @@ type Rollout interface {
 // If enough calls to DynamoDB fail, the rollout value will drop to 0 to
 // minimize possible damange (i.e. the inability to rollback a canary).
 type DynamoDBRollout struct {
-	db      *dynamodb.DynamoDB
-	table   string
-	mutex   *sync.RWMutex
-	rollout float64
+	db       *dynamodb.DynamoDB
+	table    string
+	mutex    *sync.RWMutex
+	versions map[string]version
+}
+
+func (dynamodbRollout *DynamoDBRollout) update(unhealthy time.Duration, updates map[string]float64) {
+	if updates == nil {
+		updates = make(map[string]float64)
+	}
+	dynamodbRollout.mutex.Lock()
+	// update existing entries
+	for key, val := range dynamodbRollout.versions {
+		newPercentage, ok := updates[key]
+		if !ok {
+			if time.Since(val.lastUpdated) > unhealthy {
+				if !val.isUnhealthy && val.percentage > 0 {
+					log.Printf("rollout: \"%v\" contains stale data", key)
+				}
+				val.percentage = 0
+				val.isUnhealthy = true
+			}
+		} else {
+			val.lastUpdated = time.Now()
+			val.isUnhealthy = false
+			val.percentage = newPercentage
+		}
+	}
+	// add missing entries
+	for key, val := range updates {
+		_, ok := dynamodbRollout.versions[key]
+		if !ok {
+			dynamodbRollout.versions[key] = version{
+				lastUpdated: time.Now(),
+				isUnhealthy: false,
+				percentage:  val,
+			}
+		}
+	}
+	dynamodbRollout.mutex.Unlock()
```

Comment:
Nope.  It's up to the caller(s) how they want to handle the different rollout levels.  It could be through masking (version5 20% + version6 60% = version6 60%) or through adding a trump (version6 40% + trump 20% = version6 20% trump 20%) or through any other crazy method we can't think of.  The follow-up code review to this adds such a trump.